cmd/argo/commands/root.go

```diff
@@ -133,5 +138,22 @@ If your server is behind an ingress with a path (you'll be running "argo server
 	command.PersistentFlags().IntVar(&glogLevel, "gloglevel", 0, "Set the glog logging level")
 	command.PersistentFlags().BoolVarP(&verbose, "verbose", "v", false, "Enabled verbose logging, i.e. --loglevel debug")
 
+	viper.AutomaticEnv()
+	viper.SetEnvPrefix("ARGO")
+	viper.SetEnvKeyReplacer(strings.NewReplacer("-", "_", ".", "_"))
+	// bind flags to env vars (https://github.com/spf13/viper/tree/v1.17.0#working-with-flags)
```

Comment:
hate to be anal but was thinking we could clarify in a comment that we're enabling somebody to use environment variables instead of flags, prefixed by "ARGO" with character replacement

I know there's an idea that maybe sometimes code can kind of explain itself. But I just think most of these functions people have never seen before, and it would be nice to know what this whole block of code is doing without having to read through each line and look up the function definition. Let me know if you disagree.

Reply:
mmm I do think the first 3 lines are pretty self-explanatory, which is why I only added comments to the somewhat opaque and fairly specific workarounds below those lines. with this new comment, I think that inference on the first 3 lines becomes even more simple (as env vars are directly mentioned).

also if a contributor is familiar with Argo's [environment variables](https://argoproj.github.io/argo-workflows/environment-variables/), then this lines up pretty neatly when looking at the code.

if a contributor is going to make CLI modifications, then they will likely need at least some cursory knowledge of `cobra`, since the CLI is fairly coupled to `cobra`'s implementation. `pflag` and `viper` are commonly used with `cobra` as well. `viper` is pretty commonly used for config files and/or env vars.

and more generally speaking, there's a bit of a slipper slope there -- why not comment every line of usage of `cobra` in the codebase then? I think it is reasonable that if someone wants a deep understanding or wants to modify a specific functionality, such as env vars in this case, that they will need to have some familiarity with the underlying library that handles that.

for a point of reference, I had actually never written code for a Go CLI before working on Argo. I had seen many Go CLIs use `cobra` though and had heard of `viper` before (though not seen much code with it). When I first saw the `viper` code I immediately thought "ah this must be the env var -> flag mapping". I then checked the `blame` to confirm that and did not look up any of these functions. The `BindPFlags` and `VisitAll` blocks vaguely made sense to me as well at a glance.
I only read through `viper` in more detail for this PR specifically.
(you might say I am perhaps an outlier when it comes to points of reference though 😅)

Reply:
I've worked in software for 2 decades but not on a CLI before. The concept of "flag to environment variable" mapping was not something I was aware of. For anyone like me, I think it should be easy to just read the comment and know what it's for.

Reply:
The new comment does specifically explain the mapping part though? Not sure if I'm misunderstanding something, but I thought you were referencing the first 3 lines that have no comment currently. The first 3 lines just set up env vars in general

Reply:
Why not make the comments more user friendly?....like: 
```
// enable environment variables to be passed instead of flags
// for each flag, an environment variable capitalized, prefixed by "ARGO", and with periods and dashes replaced with underscores
```

Reply:
I don't think that spelling out word-for-word _exactly_ what each line does sets a good precedent, but I can meet you somewhere in the middle and give an explanation off what the entire `viper` block as a whole is intended to do

Reply:
See [new commit](https://github.com/argoproj/argo-workflows/pull/12145/commits/27765a9e6a75ad55ed1ef2528db132cca9a12bf2)

Reply:
Thanks for the compromise. Do you mean the precedent of insisting on more comments from others? Is one of your concerns the maintenance of comments? That we may forget to change a comment when the code changes? I feel we need to balance that with efficient readability.

Re-reading your comments above, I think one difference in our perspectives is that you’re referring to somebody who’s modifying the code here. I am just thinking of the person who’s trying to read it and quickly get a sense of how it works.

Reply:
> Do you mean the precedent of insisting on more comments from others? Is one of your concerns the maintenance of comments?

No I mean specifically of commenting to explain what _each line_ does. As you know, I heavily document things, so I most certainly believe that comments have their place.
Explaining a confusing workaround -- absolutely, useful in pretty much every scenario I can think of as otherwise it's difficult to tell why code was written the way it was; this should be done more often and I'll say this myself in code reviews too. A summary of a large block of related code, sure, I think that can add value sometimes when quickly reading through; the comment effectively lets you skip through.

_Each line_ though, I hesitate quite a bit with that. Every single statement shouldn't be explained and how do you draw a boundary there? which function calls to which libraries should be commented and which shouldn't?
Like `SetEnvPrefix` and `SetEnvKeyReplacer` are pretty self-explanatory for me and are not confusingly named to me. They are also common use-cases of env vars in general.

> I think one difference in our perspectives is that you’re referring to somebody who’s modifying the code here. I am just thinking of the person who’s trying to read it and quickly get a sense of how it works.

I referred to both. The first time I read it I wasn't modifying the env vars and I got the gist of what the code does because I already knew that Argo allowed env vars instead of flags from [the docs](https://argoproj.github.io/argo-workflows/environment-variables/) and from having used that functionality myself before. The `viper` code was also only in 2 files before this (now 3 files), only the Controller and Server executables specifically.

The gist of it I think is enough for a reader. Knowing specifically what each function call to a library does I think falls more within the realm of a modifier, who should ideally read the library's docs to get familiar with the usage they plan to modify.

Reply:
Yeah, I agree on not commenting every line of code. In my example the comments at the top were for an entire code block. I agree they were pretty detailed. I’m good with your shorter alternative. I think I was going for extra clarity because it explained the interface to the application. I didn’t realize this is explained in the docs as well. Thanks for talking it out and good to know. I think our perspectives aren’t really that different but maybe we both different err differently.

Reply:
> I think our perspectives aren’t really that different

Indeed that seems to be the case most of the time 🙂